modin/engines/base/frame/data.py

```diff
@@ -1651,7 +1662,25 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
         """
         if isinstance(other, type(self)):
             other = [other]
-        if all(o.axes[axis].equals(self.axes[axis]) for o in other):
+
+        is_partitons_aligned = True
+        if (
+            len(self._partitions) != len(other[0]._partitions)
+            and len(self.axes[0]) == len(other[0].axes[0])
+            and axis == 0
+        ):
+            is_partitons_aligned = False
+            self._partitions = self._frame_mgr_cls.map_axis_partitions(
+                axis, self._partitions, lambda df: df
+            )
+            other[0]._partitions = other[0]._frame_mgr_cls.map_axis_partitions(
+                axis, other[0]._partitions, lambda df: df
+            )
```

Comment:
[`simple_shuffle`](https://github.com/modin-project/modin/blob/a571e10e00f3dc966f2ff001b2ba4dee322b95e9/modin/engines/base/frame/data.py#L1688-L1718) seems to do the same

Reply:
`simple_shuffle` does the same if indices of `self` and `other` are equal. However, there is possible cases where the indices are not equal. See example from test ([link](https://github.com/modin-project/modin/pull/2256/files#diff-2304006b7bb0a5771b48f73d4397bf59c98fe2395af347687bc803840f5f168eR223)).

Reply:
I looked into the problem a little bit deeper. Perhaps, we could avoid these additional changes and find out the problem from where it is coming. Namely, from [here](https://github.com/modin-project/modin/pull/2256/files#diff-64aee86078c1a929c18582ab58f1bc06e251bf21fbf10f523ce433f56ab7b8b0R1700). Can you investigate why we can't reindex df?

Reply:
@YarShev We can't reindex df if we have duplicate axis values. This happens after `append`, for example.

Reply:
Yes, looks like it is so.

Reply:
Then, there is the following issue:
```
import modin.pandas as pd
import pandas
data = [0, 1, 2, 3, 4, 5]
mdf1 =pd.DataFrame({"a": data, "b": data}, index=[0, 1, 2, 0, 1, 2])
mdf2 = pd.DataFrame({"a": data, "b": data})
mdf1 / mdf2
ValueError: cannot reindex from a duplicate axis
pdf1 = pandas.DataFrame({"a": data, "b": data}, index=[0, 1, 2, 0, 1, 2])
pdf2 = pandas.DataFrame({"a": data, "b": data})
pdf1 / pdf2
     a    b
0  NaN  NaN
0  inf  inf
1  1.0  1.0
1  4.0  4.0
2  1.0  1.0
2  2.5  2.5
3  NaN  NaN
4  NaN  NaN
5  NaN  NaN
```

Reply:
@YarShev thanks, this is another one additional bug. It was fixed and [tested](https://github.com/modin-project/modin/pull/2256/files#diff-2304006b7bb0a5771b48f73d4397bf59c98fe2395af347687bc803840f5f168eR245-R251).